File: solutions/3090-maximum-length-substring-with-two-occurrences/sliding-window.cpp
// time: O(n)
// space: O(1)

#include <algorithm>

using namespace std;

class Solution {
  public:
    int maximumLengthSubstring(string s) {
        int freq[26] = {};
        const int n = s.length();
        int left = 0;
        int right = 0;
        int result = 0;

        for (; right < n; right++) {
            const char ch = s[right];
            const int idx = ch - 'a';
            freq[idx]++;

            while (freq[idx] > 2) {
                const char ch = s[left];
                const int idx = ch - 'a';
                freq[idx]--;
                left++;
            }

            result = max(result, right - left + 1);
        }

        return result;
    }
};

File: solutions/3090-maximum-length-substring-with-two-occurrences/sliding-window.js
// time: O(n)
// space: O(1)

/**
 * @param {string} s
 * @return {number}
 */
const maximumLengthSubstring = s => {
    const freq = new Array(26).fill(0);
    const n = s.length;
    const offset = 97;
    let result = 0;
    let right = 0;
    let left = 0;

    for (; right < n; right++) {
        const char = s[right];
        const code = char.charCodeAt(0);
        const idx = code - offset;
        freq[idx]++;

        while (freq[idx] > 2) {
            const char = s[left];
            const code = char.charCodeAt(0);
            const idx = code - offset;
            freq[idx]--;
            left++;
        }

        result = Math.max(result, right - left + 1);
    }

    return result;
};
